### PR TITLE
task/HW-61254-iOS-Add-clearInstance-in-Core-SDK,-UI-SDK-and-Insights-SDK

### PR DESCRIPTION
**Pull request by asenthil-altmtrk**

### CODE REVIEW

**Review by jsaini-hw — COMMENTED** [review]
Move the PR to merge to development branch instead of master directly

File: Sources/Insights.swift [hunk 1]
```diff
@@ -168,7 +174,7 @@ public final class Insights: InsightsProtocol {
         }
     }
 
-    private func addObserversForTrackingLifecycleEvents() {
+    private static func addObserversForTrackingLifecycleEvents() {
```
Comment (manikandan-altmtrk) [inline Sources/Insights.swift]: May i know why you changed to static?

File: Sources/Insights.swift [hunk 2]
```diff
@@ -180,7 +186,7 @@ public final class Insights: InsightsProtocol {
                                                object: application)
     }
 
-    private func removeObserversForTrackingLifecycleEvents() {
+    private static func removeObserversForTrackingLifecycleEvents() {
```
Comment (manikandan-altmtrk) [inline Sources/Insights.swift]: May i know why you changed to static?

File: Tests/InsightsTests.swift [hunk 3]
```diff
@@ -13,6 +13,7 @@ final class InsightsTests: XCTestCase {
     override func setUp() {
         super.setUp()
         EventManager.shared.deleteEvents(before: Date().epochMilliseconds())
+        Insights.clearInstance()
```
Comment (nirmal-altmtrk) [inline Tests/InsightsTests.swift]: Add the separate function for clearInstance as same as in Core SDK.

File: Sources/Insights.swift [hunk 4]
```diff
@@ -106,9 +106,13 @@ public final class Insights: InsightsProtocol {
         self.userToken = userToken
         visitId = UUID().uuidString
     }
-
-    deinit {
-        removeObserversForTrackingLifecycleEvents()
+    /// Clears Insights instance.
```
Comment (nirmal-altmtrk) [inline Sources/Insights.swift:110]: Add a new line here.

File: Tests/InsightsTests.swift (hunk 3 above)
Comment (nirmal-altmtrk) [inline Tests/InsightsTests.swift]: Add a separate function for clearInstance and remove this line.

    
    func testClearInstance() {
        Insights.setup(environment: environment,
                       programToken: programToken,
                       sdkVersion: sdkVersion,
                       apiUrl: InsightsTestHelper.restURL,
                       userToken: userToken)
        XCTAssertNotNil(Insights.shared, "Shared instance shouldn't equal to Nil")
        Insights.clearInstance()
        XCTAssertNil(Insights.shared, "Insight instance should equal to Nil")
    }

File: Sources/Insights.swift (hunk 1 above)
Comment (asenthil-altmtrk) [reply Sources/Insights.swift]: Changed to instance method

File: Sources/Insights.swift (hunk 2 above)
Comment (asenthil-altmtrk) [reply Sources/Insights.swift]: Changed to instance method

File: Sources/Insights.swift (hunk 4 above)
Comment (asenthil-altmtrk) [reply Sources/Insights.swift:110]: Added new line

File: Tests/InsightsTests.swift (hunk 3 above)
Comment (asenthil-altmtrk) [reply Tests/InsightsTests.swift]: Added new testcase

**Review by jsaini-hw — APPROVED** [review]

File: Tests/InsightsTests.swift (hunk 3 above)
Comment (nirmal-altmtrk) [inline Tests/InsightsTests.swift]: Please remove this as since we added a separate test for clearInstance.

Comment (asenthil-altmtrk) [reply Tests/InsightsTests.swift]: Removed Insights.clearInstance()

**Review by nirmal-altmtrk — APPROVED** [review]
LGTM

**Pull request merged**